sumcheck_macro/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+                };
+                if acc.is_empty() {
+                    quote! {#arg}
+                } else {
+                    quote! {#acc, #arg}
+                }
+            },
+        );
+
+        let mut arm_body = items.iter().fold(
+            proc_macro2::TokenStream::new(),
+            |acc, (i, field_type, ident)| {
+                let f = &f_var_names[*i - 1];
+                let mut code = quote! {
+                    #acc
+                    let #ident = if let Some((start, offset)) = #f.evaluations_range() {
+                        #ident[start..][..offset].to_vec()
+                    } else {
+                        #ident[..].to_vec()
+                    };
+                };
+                if *field_type == 0 {
+                    code = quote! {
+                        #code
+                        let #ident = #ident.into_iter().map(E::from).collect::<Vec<_>>();
+                    };
+                }
+                code
+            },
+        );
+        let args = items
+            .iter()
+            .fold(proc_macro2::TokenStream::new(), |acc, (_, _, ident)| {
+                if acc.is_empty() {
+                    quote! {#ident}
+                } else {
+                    quote! {#acc, #ident}
+                }
+            });
+        arm_body = quote! {
+            #arm_body
+            (#args)
+        };
+
+        match_arms = quote! {
+            #match_arms
+            (#arm_args) => {#arm_body},
+        };
+    }
+    out = quote! {
+        #out
+
+        let (#f_tuple) = match (#match_input) {
+            #match_arms
+            _ => unreachable!(),
+        };
+    };
+
+    // Generate AdditiveArray based on degree
+    let mut additive_array_items = proc_macro2::TokenStream::new();
+    for i in 1..=(degree + 1) {
+        let single = |f: Ident| match i {
+            1 => quote! {#f[b]},
+            2 => quote! {#f[b + 1]},
+            n => join_expr(
+                quote! {+},
+                false,
+                std::iter::repeat_n(
+                    {
+                        // c1
+                        quote! {#f[b + 1] - #f[b]}
+                    },
+                    (n - 2) as usize,
+                )
+                .chain(std::iter::once(quote! {#f[b + 1]}))
+                .collect(),
+            ),
+        };
+
+        let item = join_expr(
+            quote! {*},
+            true,
+            (1..=degree)
+                .map(|j| single(ident(format!("f{j}"))))
+                .collect(),
+        );
+
+        if i == 1 {
+            additive_array_items = quote! {#item};
+        } else {
+            additive_array_items = quote! {#additive_array_items, #item};
+        }
+    }
+
+    let additive_array_items = quote! {AdditiveArray([#additive_array_items])};
+    let additive_array_first_item =
+        f_var_names
+            .iter()
+            .fold(proc_macro2::TokenStream::new(), |acc, f| {
```

Comment:
You repeat this snippet of code over and over again.  Just write a function that does it for you, and use that everywhere.